### PR TITLE
Pulled the zlib compression code into a library

### PR DESCRIPTION
**Pull request by adam-yang**
Refactor the zlib compression code into its own library. This does not change functionality, just a refactor to prepare for further work. No new tests needed.

### CODE REVIEW

**Review by tex3d — APPROVED** [review]
LGTM!

File: include/dxc/DxilCompression/DxilCompression.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    const void *pCompressedBuffer,
+    size_t BufferSizeInBytes,
+    void *pUncompressedBuffer,
+    size_t UncompressedBufferSize);
+
+  typedef void *ZlibAllocateBufferFn(void *pUserData, size_t RequiredSize);
+  ZlibResult ZlibCompress(IMalloc *pMalloc,
+    const void *pData, size_t pDataSize,
+    void *pUserData,
+    ZlibAllocateBufferFn *AllocateBufferFn,
```
Comment (dmpots) [inline include/dxc/DxilCompression/DxilCompression.h:35]: Why do we have a separate allocation function? Why is it not using the already passed IMalloc or a different IMalloc?

File: lib/DxilCompression/DxilCompression.cpp [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+#include "dxc/Support/Global.h"
+#include "dxc/Support/WinIncludes.h"
+
+#include "miniz.h"
+typedef size_t ZlibSize_t;
+
+namespace {
+//
+// A resource managment class for a zlib stream.
+// Class the appropriate init and end routines on the stream.
```
Comment (dmpots) [inline lib/DxilCompression/DxilCompression.cpp:24]: This sentence does not make sense as written.

File: include/dxc/DxilCompression/DxilCompression.h (hunk 1 above)
Comment (adam-yang) [reply include/dxc/DxilCompression/DxilCompression.h:35]: It's not really an allocation function, more of a "resize buffer" callback function. Caller code might be writing into a container type like std::vector. I'll rename it and add an explanation in the comments.

Comment (dmpots) [reply include/dxc/DxilCompression/DxilCompression.h:35]: Ok, that makes sense. A comment with some explanation would be helpful!

**Review by dmpots — APPROVED** [review]

**Comment by Keenuts** [comment]
Hi Adam, thanks for refactoring!
Is it in your plans to remove the old miniz implementation in DXC (lib/Miniz, include/miniz) ?

**Pull request merged**